### DISCUSSION TITLE
Deployment vs. Statefulset

### DISCUSSION THREAD

**Discussion by ioemat** [keycloak/keycloak#11763] [Archive]
Hi keycloak community,

the keycloak operator currently creates the workloads as _deployment_  ([see](https://github.com/keycloak/keycloak/blob/main/operator/src/main/resources/base-keycloak-deployment.yaml)) which means that the pods are replaced by a new ones when they are restarted. 

In my understanding JGroups Cluster is designed to work with stable node identities like they are provided by _statefulset_.
Using a _statefulset_ also ensures that pods are created sequentially, which probably lowers the risk of race conditions for the initial startup.

What are the advantages of _deployment_, why has this been chosen for the operator?

**Comment by andreaTP** [comment]
Hi @ioemat ,

thanks for your interest!
We spoke with the Infinispan Team and assessed that the guarantees given by the Statefulset are redundant and not strictly needed (given a correct overall configuration).
We decided to go with the principle of "least power" and used a plain "Deployment".

Additionally, and as a proof, we developed a detailed testing scenario:
https://github.com/keycloak/keycloak/blob/e2cf6ae92b1dbd91089de4da7a721e5222b3edb9/operator/src/test/java/org/keycloak/operator/ClusteringE2EIT.java#L94
to validate that the setup is working as expected.

cc. @vmuzikar

**Reply by sschu** [reply]
However, there are definitely some downsides to using deployments, first and foremost that the random pod names make handling the pods and analysing issues much much harder than the predictable ordered names that stateful sets provide. You have a first pod, a second pod and a third pod to talk about when analysing issues and discussing them with somebody else. 
Every monitoring dashboard that you have profits from stable pod names as you can see the history of the "same" pods over an extended period of time. With pods getting assigned new names all the times dashboards become overcrowded with graphs. 
Last but not least, Keycloak essentially is a stateful application with sessions etc. stored in memory, so even if it "works" with a deployment, it is not the natural fit and might lead to problems downstream. 
Given the disadvantages of Deployments I lined out, I would really ask you to reconsider this choice. We have been using Keycloak with Stateful Sets with more than 60 installations per Kubernetes cluster for years now successfully and I don't really see an advantage going to a deployment, only disadvantages.
@andreaTP @vmuzikar @stianst

**Reply by andreaTP** [reply]
Hi @sschu , and thanks for chipping in this discussion!

> first and foremost that the random pod names make handling the pods and analysing issues much much harder

this is arguably not correct as long as the pods are properly labeled, and every tool I'm aware of [supports `matchLabels`](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#api) out-of-the-box.

> Every monitoring dashboard that you have profits from stable pod names as you can see the history of the "same" pods over an extended period of time. 

this is not "profit" in my opinion, but simple aggregation based on an arguable and arbitrary criterion, since on restart the pod is not the same anymore even if it has the "same name".

> With pods getting assigned new names all the times dashboards become overcrowded with graphs.

This issue can, and should, be tackled in different ways with retention periods/aggregating information etc. etc.

> Last but not least, Keycloak essentially is a stateful application with sessions etc. stored in memory, so even if it "works" with a deployment, it is not the natural fit and might lead to problems downstream.

Keycloak is supposed to work in a cluster leveraging the (for now)embedded Infinispan, this configuration is extensively tested and used in production scenarios.
Part of the Store.X effort is to make Keycloak even more stateless and using a `Deployment` nicely aligns with the current situation and the plans for the future.

**Reply by sschu** [reply]
> this is arguably not correct as long as the pods are properly labeled, and every tool I'm aware of [supports matchLabels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#api) out-of-the-box.

I was talking about analysing issues like, pod b restarted, pod a became coordinator, pod a transferred all session data to pod b and ran into an out of memory exceptions. Issues you might have to analyse because Keycloak pods interact with each other compared to traditional 12 factor apps that put all the difficult stateful stuff into some backends.

> This issue can, and should, be tackled in different ways with retention periods/aggregating information etc. etc.

So you are telling me I have to do a lot of extra work to get something I am getting for free with Stateful Sets? ;)

> Part of the Store.X effort is to make Keycloak even more stateless and using a Deployment nicely aligns with the current situation and the plans for the future.

Once Keycloak is truly stateless (which I doubt ever happens for good reasons), I am fine with a Deployment. So far I can only see disadvantages and extra work I have to do.

**Reply by andreaTP** [reply]
> I was talking about analysing issues like, pod b restarted, pod a became coordinator, pod a transferred all session data to pod b and ran into an out of memory exceptions. Issues you might have to analyse because Keycloak pods interact with each other compared to traditional 12 factor apps that put all the difficult stateful stuff into some backends.

I fail to understand how having unique names for pods makes those analyses harder rather than simpler. If a pod gets restarted you have immediate evidence due to the new name instead of relying on more aggregated metrics.

> So you are telling me I have to do a lot of extra work to get something I am getting for free with Stateful Sets? ;)

Giving more work to do to our users is never going to be the objective.
Nevertheless, a "StatefulSet" implies more guarantees from the runtime that is unnecessary, and, following [the principle of least power](https://en.wikipedia.org/wiki/Rule_of_least_power) we are using the less powerful "Deployment" to ease user's operations.

**Reply by rh-iatanaso** [reply]
This is a great discussion and I wanted to chime in. Disclosure:  currently our keycloak deployments  are running Stateful Sets and they are the exception everything else we run is using Deployments.
@andreaTP I do disagree a bit with 
> the guarantees given by the Stateful Set are redundant and not strictly needed

One example where the Stateful Sets are needed at the current state in my opinion is:
* In bigger clusters with distributed caches, those are spread between some nodes. The number of nodes must be relatively low because  the number of replication impacts the cluster stability at the current state. We do want to be able to have release often without downtime. Loosing cache partition forces users to login ( This is not a good user experience and affects the availability). 
* This one is more controversial, Stateful Sets have bit better protection against deployments gone bad out of the box. Single pod is rolled out and if that fails the rollout is stopped and the pods needs to be fixed before continuing. This is a bandaid but it is useful because currently infinispan more or less keep state in some cases.
* Arguably one can probably create **Deployment** that deploys pod one at a time but the **Stateful Set** provides that out of the box. It seems that  Deployment might end up trying to replicate the Stateful Set behavior .

I think many of the points I mentioned above depend on keycloak keeping state in infinispan. If that is not an issue which is the plan I believe, Deployment will be preferred. Of course there is a chance I am completely wrong a my assessment too.

**Reply by andreaTP** [reply]
Hi @rh-iatanaso and thanks a lot for bringing those arguments to the attention!

A general observation that we forgot to mention is that a `StatefulSet` is commonly used when a pod requires a stateful backend(e.g. disk) in the form of a PVC, in the case of Keycloak(as of today) this is not the case as the database is externally provisioned.

Regarding your specific observations, I would say that configuring `podDisruptionBudget`, `nodeAffinity`, etc. accordingly you "should" be able to end up with an installation that is both resilient and can, possibly, scale better than using a `StatefulSet`.

I agree that this discussion is strongly biased(for me) by the fact that we plan to remove the embedded Infinispan and give the opportunity to deploy it standalone (e.g. externally from Keycloak).

Something that we haven't explored so far, is how much it will cost in terms of maintenance to push this decision to the users (at least until Infinispan is embedded) in the form of a field in the CRD.

**Reply by pedroigor** [reply]
@andreaTP Is it a good practice to configure node affinity to pods? I always thought it is a bad practice and you should avoid as much as possible influencing the scheduler's decisions.

**Comment by sschu** [comment]
@andreaTP What do you think about making this configurable?

**Reply by andreaTP** [reply]
If there is enough interest I can try to POC it and we can go from there

**Reply by sschu** [reply]
@andreaTP I would totally appreciate that. We could probably also support.

**Reply by vmuzikar** [reply]
IMO the Operator should be opinionated, so I don't think it should be configurable by the user. I tend to lean more towards the Deployment instead of SS as we should rather move KC towards stateless mindset (with Infinispan being external in the future) but I can imagine if there are solid reasons for SS in the current state, I'd go this path for now.

**Reply by sschu** [reply]
@vmuzikar But you could also be opinionated by providing a solid default with the ability to override?

**Reply by vmuzikar** [reply]
My reasoning here is that if we decided to make it configurable, it'd add additional complexity and maintenance costs to the Operator. Not sure if the benefit of having it configurable would justify this.

As I mentioned, maybe we should rather consider switching to SS if there are strong enough reasons for it.

**Reply by andreaTP** [reply]
> you could also be opinionated by providing a solid default with the ability to override

I think this is extremely correct, and, in this case, the complexity of the code to achieve it, is not huge.

**Comment by andreaTP** [comment]
Follow up, I quickly hacked around a version of the operator capable of switching between `Deployment` and `StatefulSet` using an environment variable in the operator `Deployment`: https://github.com/keycloak/keycloak/pull/12447

I would be happy to hear feedback on this!

**Reply by andreaTP** [reply]
cc. @ahus1 FYI you might be interested into trying this out

**Reply by sschu** [reply]
Wow, that was quick!

**Comment by pedroigor** [comment]
@andreaTP Infinispan operator seems to be using stateful set https://infinispan.org/docs/infinispan-operator/main/operator.html. Are there differences when running DG and embedding that imposes a specific workload object?

Also, once we support zero-downtime upgrades, isn't ordered and automatic rolling upgrades a must?

**Reply by ryanemerson** [reply]
@pedroigor Infinispan dev here :slightly_smiling_face:

The Infinispan operator uses a StatefulSet for a number of reasons, mainly it:

1. Allows pod data to be persisted and reloaded on pod restart. This allows for global-state and persisted data to be maintained across cluster restarts.
2. Ensures that the cluster is only scaled up/down by one pod at a time. This is important to ensure that the cluster is not scaled down too quickly, resulting in data loss if a rebalance cannot complete before `num_owners` pods are stopped.

The Infinispan Operator has to cater for a wide-range of different use-cases/workloads and using StatefulSets was the best fit for the majority.

My understanding of the keycloak use-case is that:
- 1 is not applicable, as data is not persisted and global-state is not enabled by default in Infinispan Embedded. 
- 2 is not an issue as scaling will be handled by the Keycloak Operator directly. This behaviour can also be controlled using a combination of `maxUnavailable` and `maxSurge` configurations on the Deployment. Finally, loss of cache data is not catastrophic and can be recovered from as a user will just need to obtain a new token in the worst-case scenario.

So from an data/Infinispan POV I think you're fine with a StatefulSet or Deployment. I'll let others way in on the Pros/Cons of either approach due to other considerations :smile:

**Reply by sschu** [reply]
Loss of cache data is not catastrophic, but it most likely means users have to relogin since their sessions are lost. Our customers would definitely complain about that...

**Reply by ahus1** [reply]
+1 for using a statefulset (at least as long as Infinispan is included in Keycloak) to prevent lost session information on restarting instances. 

This could happen on scaling down and on redeployments.

I did a test of scaling from 1 instance to 10 instances of Keycloak running in clustered mode with the standard nightly build of the new operator: I see messages like "no physical address", so also scaling up seems to have some problems with the deployment. Here the detailed logs: 

> {"timestamp":"2022-06-15T07:24:14.858Z","sequence":8428,"loggerClassName":"org.infinispan.util.logging.Log_$logger","loggerName":"org.infinispan.CLUSTER","level":"INFO","message":"ISPN000094: Received new cluster view for channel ISPN: [keycloak-84b48bc8dc-h9dms-15950|16] (7) [keycloak-84b48bc8dc-h9dms-15950, keycloak-84b48bc8dc-pmprq-63712, keycloak-84b48bc8dc-pxjdq-42791, keycloak-84b48bc8dc-lq5ds-50869, keycloak-84b48bc8dc-rtsjb-42104, keycloak-84b48bc8dc-5pj7r-54145, b4432256-a5bd-4023-9992-6da854e9e973]","threadName":"jgroups-17,keycloak-84b48bc8dc-5pj7r-54145","threadId":38,"mdc":{},"ndc":"","hostName":"keycloak-84b48bc8dc-5pj7r","processName":"QuarkusEntryPoint","processId":1}
> {"timestamp":"2022-06-15T07:24:14.871Z","sequence":8429,"loggerClassName":"org.jboss.logging.Logger","loggerName":"org.infinispan.CLUSTER","level":"INFO","message":"ISPN100000: Node b4432256-a5bd-4023-9992-6da854e9e973 joined the cluster","threadName":"jgroups-17,keycloak-84b48bc8dc-5pj7r-54145","threadId":38,"mdc":{},"ndc":"","hostName":"keycloak-84b48bc8dc-5pj7r","processName":"QuarkusEntryPoint","processId":1}
> {"timestamp":"2022-06-15T07:24:16.98Z","sequence":8430,"loggerClassName":"org.slf4j.impl.Slf4jLogger","loggerName":"org.jgroups.protocols.TCP","level":"WARN","message":"JGRP000032: keycloak-84b48bc8dc-5pj7r-54145: no physical address for b4432256-a5bd-4023-9992-6da854e9e973, dropping message","threadName":"TQ-Bundler-6,keycloak-84b48bc8dc-5pj7r-54145","threadId":26,"mdc":{},"ndc":"","hostName":"keycloak-84b48bc8dc-5pj7r","processName":"QuarkusEntryPoint","processId":1}

Those warnings seemed to stall the full start of the node. Could that work better with the statefulset?

**Reply by andreaTP** [reply]
We discussed the eventuality of "data loss" in case of actual disruption, given @ryanemerson answer unless we:

 - deploy using `StatefulSet`
 - enable Infinispan global-state
 - actually persist the global-state mounting PVCs accordingly
 
This is going to happen anyhow.
Since the fact that a `StatefulSet` and the `PV`/`PVC`s are huge infrastructure requests for the users, and given the fact that this is supposed to be a "worst-case scenario" situation (i.e. it might happen only in case of real disruption) I'm not going to consider this as a straight argument in favor of `StatefulSet`.

@ahus1 regarding:
> have some problems with the deployment

let's separately discuss it, gut feeling is that this is not related, you can anyhow test out the `StatefulSet` using this branch:
https://github.com/keycloak/keycloak/pull/12447

**Reply by ahus1** [reply]
I'll give the branch a try. 

To reply to the comment: I agree that Keycloak should not use a PV/PVC as those are a huge infrastructure request. I don't see the impact of a StatefulSet without PV/PVC, yet.

**Reply by kkcmadhu** [reply]
> ut it most likely means users have to relogin since their sessions are lost. Our customers would definitely complain about that...

yup. I too think the same, it does translate to bad user experience ( besies its the core use case of keycloak)

**Reply by andreaTP** [reply]
I have to stress that, losing the sessions:
 - doesn't depend *only* on switching to `StatefulSet`
 - might happen only in case of real disruption in the cluster

**Comment by pedroigor** [comment]
Thanks, @ryanemerson. 

Doesn't the #2 point (scaling up/down by one pod at a time) above also apply to embedded? I mean, considering that in the embedded mode we are also relying on distributable caches, scaling down too quickly might still result in data loss if a rebalance cannot complete before `num_owners pods` are stopped, right?

You're right that data loss is not so catastrophic but it has a direct impact on UX and user-facing availability. I'm not going to mention the impact on upgrades because that is something we are improving with the new store and there we won't have an embedded Infinispan anymore.

I'm not opinionated about SS or Deployment and I agree we need feedback from others and take into account the feedback from them when managing a production-grade cluster.

One important point you have though is that the operator should be able to solve some of the issues with using a deployment because in the end, it will be managing the workload. So perhaps using a deployment gives more flexibility to the operator in terms of the best approach for us for scaling.

**Comment by andreaTP** [comment]
I strongly appreciate everyone's feedback!
But, given that this Discussion is starting to be a little chatty and hard to follow, let me try to summarize and focus it back.

**NOTE:**
This discussion is **NOT** about "preferences" we are looking for actual **technical requirements** and proof to decide if it's best or not to change the current `Deployment` to be a `StatefulSet` as it was with the old operator.

Key takeaways(so far):

 - Embedded Infinispan is the only component inside Keycloak that can benefit from a change `Deployment`/`StatefulSet`
 - Infinispan devs agree that with the current setup/usage a `Deployment` is fine
 - `StatefulSet` comes with higher requirements than `Deployment` and we want to keep those as low as possible (i.e. [Rule of least power](https://en.wikipedia.org/wiki/Rule_of_least_power) )
 
As an additional note, I want to share that the work done on Store.X will enable Keycloak to use an external Infinispan installation lowering, even more, the requirements for `StatefulSet`s.

Last but not least, I want to remember that we are talking about the "opinionated default of the Keycloak operator" and there is absolutely no artificial barrier preventing from deploying Keycloak using `StatefulSet`s.

**Reply by pedroigor** [reply]
Considering the long-term plan toward a stateless deployment, I agree a deployment makes sense.

And I also agree that we should be considering statefulset if we find out it is more suitable for the legacy store and during the period the legacy store is not completely removed. If a deployment can work with both stores, even better.

**Reply by sschu** [reply]
@andreaTP Correction on your point 2 takeaway. They agree that it is fine as long as you are able to have only one pod join and leave at a time and let Infinispan finish its rebalance.. This is something a stateful set does out of the box. It might be possible to get the same effect using a deployment with `maxsurge=1`and `maxunavailable=0`. However, I am not sure what effect this has when a deployment is scaled down by multiple pods at the same time, if they are actually terminated one after another or at the same time. The latter might cause data loss.

**Reply by andreaTP** [reply]
@sschu thanks for the comment!
Yes, this is an action point that we should follow up on, created #12558 to track it down.
With the correction that `maxUnavailable=1` which is the [same default as a `StatefulSet`](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#maximum-unavailable-pods).

**Reply by sschu** [reply]
As I have written in https://github.com/keycloak/keycloak/pull/12447#issuecomment-1161317869, the operator would also have to apply scaling changes incrementally to the deployment to get a safe behaviour similar to a stateful set. I am not sure it makes sense to do this yourself.

**Reply by andreaTP** [reply]
I have tested and I can confirm that the settings `maxUnavailable` and `maxSurge` are applied only during a `RollingUpdate` but not during `Scaling` operations.

This is a compelling argument that prevents us from fulfilling Infinispan requirements using a plain `Deployment` ( cc. @ryanemerson @rigazilla for future reference)

**Comment by andreaTP** [comment]
We finally reached the bottom line of this thread.
There is a technical requirement of supporting pods leaving one by one the embedded Infinispan cluster and this can only be fulfilled by switching to a `StatefulSet`.

You can track the development of this change here: [#12700](https://github.com/keycloak/keycloak/issues/12700)

**Reply by paul-civitas** [reply]
Hello. What exactly are the consequences though? If you use a deployment instead of a statefulset and you have multiple pods leaving the cluster at once, what negative impact does that have?

**Reply by shawkins** [reply]
@paul-civitas see the doc note https://github.com/keycloak/keycloak/blob/22f9d2077e6842e402a7db457964a8adf2c84c99/docs/guides/getting-started/getting-started-scaling-and-tuning.adoc?plain=1#L50 - scaling down too aggressively can be problematic given the replication factor, and there have been several issues noted with members joining simultaneously - however the infinispan team is working towards addressing that. So for now it's been easiest to just use a StatefulSet. In the future we may be able to use Deployments and pod disruption budgets.

**Reply by paul-civitas** [reply]
@shawkins yes but which data specifically can be lost this way? Dataloss isn't always a disaster depending on the data. In particular for cache data loss, the only consequence is cache misses. What are the several issues that have been noted? As you scale up it gets harder and harder to avoid multiple members joining or leaving concurrently.

**Reply by shawkins** [reply]
> yes but which data specifically can be lost this way?

Session data. This should be a non-issue once persistent sessions are commonly in use.

>  What are the several issues that have been noted?

https://issues.redhat.com/browse/ISPN-15432
https://issues.redhat.com/browse/ISPN-15586
https://issues.redhat.com/browse/ISPN-15872

There's also a concern about how expensive rebalancing can be - there again that's something persistent sessions may be able eventually address, but I'm not sure if it's there yet.

Since this discussion is quite old, if you want to revisit this based upon specific concerns that would be relevant to the planning for KC 27 and later, it would be best to open a new discussion.

**Reply by pruivo** [reply]
In Keycloak 26, sessions are stored in the database and should be safe. However, some data is only stored in the caches:

* Invalidation events to invalidate caches, when user/realm information changes, an event is stored in the cache. If the event is lost, Keycloak pods may keep outdated cached information indefinitely. 
* Single-use tokens may be lost or worse, they may be used but this information is not replicated
* Authentication session loss can lead to unnecessary user session creation leading to increased resource usage (and may annoy users).

Stateful sets reduce the risks above while deployment increases them. 

Improving the Infinispan handling of concurrent nodes leaving/joining is an outgoing process. Unfortunately, we are still finding issues with it: https://issues.redhat.com/browse/ISPN-16531

**Comment by imeliran** [comment]
Hi everyone, thanks for the insightful discussion here!
I'm wondering if the points raised in this thread might also be relevant to [issue #39866](https://github.com/keycloak/keycloak/issues/39866). The contexts are somewhat different, but the underlying problem space seems related.

**Reply by ahus1** [reply]
Yes, they are related to that issue. We still want to have a fix for that issue that concurrent starts work without problems. 

BTW, those watching this issue, there is now a Kubernetes stateful-set example linked in https://www.keycloak.org/getting-started/getting-started-kube
